napari/config/core.py

```diff
@@ -0,0 +1,708 @@
+the file and the config have changed, it prioritizes values in the config.
+"""
+import ast
+import logging
+import os
+import sys
+import threading
+from collections import defaultdict
+from functools import wraps
+from types import TracebackType
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    Union,
+)
+
+import yaml
+
+from ..utils.misc import StringEnum
+from .deprecations import check_deprecations
+from .utils import canonical_name, ensure_file, merge, update
+
+logger = logging.getLogger('napari.config')
+
+
+#: a list of paths that get searched for .yaml files in :func:`collect_yaml`
+paths = [
+    os.getenv("NAPARI_ROOT_CONFIG", "/etc/napari"),
+    os.path.join(sys.prefix, "etc", "napari"),
+    os.path.join(os.path.expanduser("~"), ".config", "napari"),
+    os.path.join(os.path.expanduser("~"), ".napari"),
+]
+
+
+#: The primary directory for user config files
+#:
+#: Defaults to ~/.config/napari
+#: May be set with the environment variable "NAPARI_CONFIG"
+PATH = os.path.join(os.path.expanduser("~"), ".config", "napari")
+if "NAPARI_CONFIG" in os.environ:
+    PATH = os.environ["NAPARI_CONFIG"]
+    paths.append(PATH)
+
+#: a special yaml file to dump settings that are changed within a session
+#:
+#: this file will be loaded, but takes lower priority than all non-private
+#: yaml files.  So if a user has a specific setting in their napari.yaml file,
+#: it will override anything in _session.yaml
+_SESSION = os.path.join(PATH, '_session.yaml')
+
+
+#: the main global config dict
+config: Dict[str, Any] = {}
+
+CallbackDict = DefaultDict[Tuple[str, ...], Set[Callable[[Any], None]]]
+#: mapping from key (as tuple of strings), to callback function
+#:
+#: callback will be called whenever the key changes
+callbacks: CallbackDict = defaultdict(set)
+
+config_lock = threading.Lock()
+
+#: A list of config dicts that downstream libraries may use to register defaults
+#:
+#: see :func:`update_defaults` and
+#: https://docs.dask.org/en/latest/configuration.html#downstream-libraries
+defaults: List[dict] = []
+
+no_default = "__no_default__"
+
+
+def get(key: str, default=no_default, config: dict = config) -> Any:
+    """Get elements from global config.
+
+    Use '.' for nested access
+
+    Examples
+    --------
+    >>> from napari import config
+    >>> config.get('foo')
+    {'x': 1, 'y': 2}
+
+    >>> config.get('foo.x')
+    1
+
+    >>> config.get('foo.x.y', default=123)
+    123
+
+    See Also
+    --------
+    napari.config.set
+    """
+    keys = key.split(".")
+    result = config
+    for k in keys:
+        k = canonical_name(k, result)
+        try:
+            result = result[k]
+        except (TypeError, IndexError, KeyError):
+            if default is not no_default:
+                return default
+            else:
+                raise
```

Comment:
I could imagine case where function is called in two places with two different default values. So when value is not defined in configuration then there will be inconsistency in viewer state. I suggest to recursive create when default is provided.